src/hellogo_fares/utils.rs

```diff
@@ -0,0 +1,954 @@
+// Copyright 2017 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+use crate::hellogo_fares::TryOnlyChild;
+use crate::objects::Date;
+use crate::Result;
+use chrono::NaiveDate;
+use failure::{bail, format_err, Error};
+use minidom::Element;
+use rust_decimal::Decimal;
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+use std::str::FromStr;
+
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub enum FrameType {
+    Resource,
+    Service,
+    UnitPrice,
+    DistanceMatrix,
+    DirectPriceMatrix,
+}
+
+impl Display for FrameType {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            FrameType::Resource => write!(f, "Resource"),
+            FrameType::Service => write!(f, "Service"),
+            FrameType::UnitPrice => write!(f, "UnitPrice"),
+            FrameType::DistanceMatrix => write!(f, "DistanceMatrix"),
+            FrameType::DirectPriceMatrix => write!(f, "DirectPriceMatrix"),
+        }
+    }
+}
+
+impl FromStr for FrameType {
+    type Err = Error;
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "UnitPrice" => Ok(FrameType::UnitPrice),
+            "DistanceMatrix" => Ok(FrameType::DistanceMatrix),
+            "DirectPriceMatrix" => Ok(FrameType::DirectPriceMatrix),
+            _ => bail!("Failed to convert '{}' into a FrameType", s),
+        }
+    }
+}
+
+pub fn get_value_in_keylist<F>(element: &Element, key: &str) -> Result<F>
+where
+    F: FromStr,
+{
+    let values = element
+        .try_only_child("KeyList")?
+        .children()
+        .filter(|key_value| match key_value.try_only_child("Key") {
+            Ok(k) => k.text() == key,
+            _ => false,
+        })
+        .map(|key_value| key_value.try_only_child("Value"))
+        .collect::<Result<Vec<_>>>()?;
+    if values.len() != 1 {
+        bail!(
+            "Failed to find a unique key '{}' in '{}'",
+            key,
+            element.name()
+        )
+    } else {
```

Comment:
you can remove this line

Reply:
Done in 812eec101cf927aabfedd0b13cd85d935ebe9db2.